proto-lens-protoc/src/Data/ProtoLens/Compiler/Definitions.hs

```diff
@@ -163,60 +177,67 @@ messageAndEnumDefs protoPrefix hsPrefix messages enums
+                    }
+      where
+        n' = fieldName n
+    oneofInfo :: Int32 -> Text -> OneofInfo
+    oneofInfo idx n = OneofInfo
+                        { oneofFieldName = mkFieldName n
+                        , oneofTypeName = fromString $ hsPrefix' ++ hsName n
+                        , oneofCases = map oneofCase
+                                          $ Map.findWithDefault [] (Just idx)
+                                              allFields
+                        }
+    oneofCase f = OneofCase
+                        { caseField = fieldInfo f
+                        , caseConstructorName =
+                              -- Note: oneof case constructors aren't prefixed
+                              -- by the oneof name; field names (even inside
+                              -- of a oneof) are unique within a message.
+                              fromString $ hsPrefix' ++ hsName (f ^. name)
+                        }
 
-    thisDef = (protoPrefix <> protoName
-              , Message MessageInfo
-                  { messageName = fromString $ hsPrefix ++ hsName (d ^. name)
-                  , messageDescriptor = d
-                  , messageFields = allFields
-                  , messageOneofFields = ooInfos
-                  })
-    subDefs = messageAndEnumDefs protoPrefix' hsPrefix'
-                  (d ^. nestedType) (d ^. enumType)
-    protoPrefix' = protoPrefix <> protoName <> "."
-    hsPrefix' = hsPrefix ++ hsName (d ^. name) ++ "'"
-    extractFields p = [ fieldInfo n f
-                      | f <- (d ^. field), p f
-                      , let n = fieldName (f ^. name)
-                      ]
-    recFieldName n = fromString $ "_" ++ hsPrefix' ++ n
-    fieldInfo n descriptor = FieldInfo
-        { overloadedField = n
-        , recordFieldName = recFieldName n
-        , fieldDescriptor = descriptor
-        , oneofFieldInfo = Nothing
-        }
-    oneofInfo n idx =
-        let typename = hsPrefix' ++ hsName n
-            encFieldName = recFieldName $ fieldName n
-            oneofFields = [ info { oneofFieldInfo = Just $ OneofFieldInfo
-                                    { oneofConstructorName = fromString $ typename ++ "'" ++ overloadedField info
-                                    , oneofEnclosingFieldName = encFieldName
-                                    }
-                                 }
-                          | info <- extractFields (\f -> elem idx (f ^. maybe'oneofIndex))
-                          ]
-        in (oneofFields, OneofInfo
-               { oneofTypeName = typename
-               , oneofRecordFieldName = encFieldName
-               })
+
+-- | Group fields by the index of the oneof field that they belong to.
+-- (Or 'Nothing' if they don't belong to a oneof.)
+collectFieldsByOneofIndex
+    :: [FieldDescriptorProto] -> Map.Map (Maybe Int32) [FieldDescriptorProto]
+collectFieldsByOneofIndex =
```

Comment:
Done with `fromListWith`.  I decided to avoid `&&&/pure` since I feel the explicit lambda is a little more readable.